### PR TITLE
Fixed the issue of incorrect voltage results in circuit connections.

### PR DESCRIPTION
**Pull request by Asachoo**
Try to fix the issue https://github.com/scikit-rf/scikit-rf/issues/1106, In this PR, the output of the case mentioned in the Issue are shown as blow:
```bash
Result for z0: [[50.+0.j 50.+0.j]]:
[[[-0.21995821+1.22760113j -0.25338403-1.11897886j]
  [-0.27041083-0.0546119j   0.99939488+1.82918502j]]] S-parameters in external port
[[ 7.80041793+12.27601126j  7.80041793+12.27601126j
  -1.44312993 +1.28197898j -1.44312993 +1.28197898j
  -2.70410829 -0.546119j   -2.70410829 -0.546119j  ]] Voltage across ports

Result for z0: [[40.+0.j 40.+0.j]]:
[[[-0.21995821+1.22760113j -0.25338403-1.11897886j]
  [-0.27041083-0.0546119j   0.99939488+1.82918502j]]] S-parameters in external port
[[ 7.80041793+12.27601126j  7.80041793+12.27601126j
  -1.44312993 +1.28197898j -1.44312993 +1.28197898j
  -2.70410829 -0.546119j   -2.70410829 -0.546119j  ]] Voltage across ports
```
But I am still not sure whether the modification in this PR is completely correct, and I hope to get your suggestions. If this modification is proven to be correct, I will modify the calculation method of `Circuit.current()`.

Your feedback and suggestions on this optimization are welcome!

### CODE REVIEW

**Comment by Asachoo** [comment]
The Circuit.current() method has been modified and verified by the case:

```python
import numpy as np

import skrf as rf


def gen_random_s(npoints: int, nports: int = 2) -> np.ndarray:
    s = 2 * np.random.rand(npoints, nports, nports) * 1.0j - 1
    s += 2 * np.random.rand(npoints, nports, nports) - 1
    return s


np.random.seed(123)
npoints = 1
freq = rf.Frequency(1, 100, npoints, "ghz")

porta = rf.Circuit.Port(frequency=freq, name="porta")
portb = rf.Circuit.Port(frequency=freq, name="portb")

ntwka = rf.Network(frequency=freq, s=gen_random_s(npoints), name="ntwka")
ntwkb = rf.Network(frequency=freq, s=gen_random_s(npoints), name="ntwkb")
ntwkc = rf.Network(frequency=freq, s=gen_random_s(npoints), name="ntwkc")

ntwkx = ntwka.copy()
ntwky = ntwkb.copy()
ntwkz = ntwkc.copy()

ntwkx.renormalize(40)
ntwky.renormalize(30)
ntwkz.renormalize(20)

for ntw1, ntw2, ntw3 in (
    (ntwka, ntwkb, ntwkc),
    (ntwkx, ntwky, ntwkz),
):
    circuit = [
        [(porta, 0), (ntw1, 0)],
        [(ntw1, 1), (ntw2, 0)],
        [(ntw2, 1), (ntw3, 0)],
        [(ntw3, 1), (portb, 0)],
    ]

    circuit = rf.Circuit(circuit)
    print(f"Result for z0: {ntw1.z0[0]}, {ntw2.z0[0]}, {ntw3.z0[0]}:")
    print(circuit.s_external, "S-parameters in external port")

    print(circuit.currents([1, 0], [0, 0]), "Current across ports")
    print(circuit.voltages([1, 0], [0, 0]), "Voltage across ports")
    print()
```

Output:
```bash
Result for z0: [50.+0.j 50.+0.j], [50.+0.j 50.+0.j], [50.+0.j 50.+0.j]:
[[[-0.29645462+1.1632858j   0.20393204+0.02071413j]
  [ 0.03676862-0.12222617j -0.64527139+0.63791598j]]] S-parameters in external port
[[-0.25929092+0.23265716j  0.25929092-0.23265716j -0.06306092-0.0103588j
   0.06306092+0.0103588j   0.03483   -0.01447254j -0.03483   +0.01447254j
  -0.00735372+0.02444523j  0.00735372-0.02444523j]] Current across ports
[[ 7.03545381+11.63285797j  7.03545381+11.63285797j
  -2.11274857 +1.88695305j -2.11274857 +1.88695305j
  -0.35352955 -0.32933193j -0.35352955 -0.32933193j
   0.36768618 -1.22226169j  0.36768618 -1.22226169j]] Voltage across ports

Result for z0: [40.+0.j 40.+0.j], [30.+0.j 30.+0.j], [20.+0.j 20.+0.j]:
[[[-0.29645462+1.1632858j   0.20393204+0.02071413j]
  [ 0.03676862-0.12222617j -0.64527139+0.63791598j]]] S-parameters in external port
[[-0.25929092+0.23265716j  0.25929092-0.23265716j -0.06306092-0.0103588j
   0.06306092+0.0103588j   0.03483   -0.01447254j -0.03483   +0.01447254j
  -0.00735372+0.02444523j  0.00735372-0.02444523j]] Current across ports
[[ 7.03545381+11.63285797j  7.03545381+11.63285797j
  -2.11274857 +1.88695305j -2.11274857 +1.88695305j
  -0.35352955 -0.32933193j -0.35352955 -0.32933193j
   0.36768618 -1.22226169j  0.36768618 -1.22226169j]] Voltage across ports
```

**Review by jhillairet — APPROVED** [review]

File: skrf/circuit.py [hunk 1]
```diff
@@ -1233,7 +1237,11 @@ def voltages(self, power: NumberLike, phase: NumberLike) -> np.ndarray:
         b = self._b(a)
         z0s = self.z0
         directions = self._currents_directions
-        Vs = (b[:,directions[:,0]] + b[:,directions[:,1]])*np.sqrt(z0s)
+        i_l, i_r = directions[:,0], directions[:,1]
+
+        z0_sqrt = np.sqrt(z0s)
+        gamma = (z0s[:,i_r]-z0s[:,i_l]) / (z0s[:,i_r]+z0s[:,i_l])
+        Vs = (b[:,i_l] * z0_sqrt[:,i_r] + b[:,i_r] * z0_sqrt[:,i_l]) * np.sqrt(1 - gamma**2)
```
[the remark this answers is not in the document]
Comment (Asachoo) [reply skrf/circuit.py]: Thanks for the suggestion! I've adopted the proposed changes.

**Comment by Asachoo** [comment]
Thanks for the tips, @jhillairet. The current fix is indeed just a preliminary step; my future focus will be on adding support for complex impedances.

**Comment by jhillairet** [comment]
Ok. Complex reference impedance is a kind of hell: With the power waves reference impedance definition, allowing complex reference impedances lead to very cumbersome expressions (I've derived a few, but it's hard to simplify them much into closed forms...). And it also leads to theoretical inconsistencies.

So, while implementing complicated expressions should not be such a problem, the proper question is instead: why do we need complex reference impedances?

So far, I haven't found real life usages. Do you know some?

In addition, all the commercial circuit solvers that I know do not support complex reference impedances. So, how can we validate our implementation on advanced cases? 

I was stuck with these two issues in the past...

**Comment by Asachoo** [comment]
I completely agree with your perspective on the challenges posed by introducing complex reference impedances. As you mentioned, the practical applications for complex reference impedances are quite limited—I've only encountered them in HFSS touchstone files.

The expectation that `Circuit` supports complex reference impedances is based on the success in `Network`, which handles them flawlessly. It would indeed be regrettable if Circuit were restricted to using `Network` with only real impedances. In addition, the absence of explicit prompts or restrictions in `Circuit`'s DocString and corresponding method suggests that complex reference impedances are feasible, despite the inherent complexities.

Therefore, until the issues surrounding the expressions, application, and validation of complex reference impedances are resolved, would a transitional changes like the following be acceptable?

1. Provide a DocString in Circuit's current implementation indicating partial support for complex reference impedances.
2. Raise a NotImplementedError during current/voltage calculations when encountering complex reference impedances, similar to https://github.com/scikit-rf/scikit-rf/blob/fccd6562060f8e03c8d060eb054f4bddc54014ea/skrf/circuit.py#L1241-L1243


This approach could serve as a temporary solution while the support for complex reference impedances become true.

**Comment by jhillairet** [comment]
> I've only encountered them in HFSS touchstone files.

Yeah me too. When ports are surrounded by lossy materials or lossy boundary conditions, reference impedances are then (slightly) complex.

> Therefore, until the issues surrounding the expressions, application, and validation of complex reference impedances are resolved, would a transitional changes like the following be acceptable?

Sure, I totally agree to your proposals. And for the moment, it can be indicated that complex reference impedances are not supported.

**Review by jhillairet — APPROVED** [review]

**Comment by jhillairet** [comment]
can we merge this?

**Comment by Asachoo** [comment]
Yes, I consider it finished.

**Pull request merged**